File: src/main/java/net/insprill/cjm/messages/MessageSender.java
package net.insprill.cjm.messages;

import com.google.common.reflect.ClassPath;
import lombok.Getter;
import net.insprill.cjm.CJM;
import net.insprill.cjm.handlers.PlayerHandler;
import net.insprill.cjm.handlers.RandomHandler;
import net.insprill.cjm.handlers.VanishHandler;
import net.insprill.cjm.messages.types.MessageType;
import net.insprill.xenlib.XenLib;
import net.insprill.xenlib.XenMath;
import net.insprill.xenlib.XenUtils;
import net.insprill.xenlib.files.YamlFile;
import org.bstats.charts.SimplePie;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.event.Listener;
import org.bukkit.permissions.PermissionDefault;

import java.io.IOException;
import java.util.*;

public class MessageSender implements Listener {

    private static MessageSender instance;

    public static MessageSender getInstance() {
        return instance;
    }

    @Getter
    private final Map<String, MessageType> messageTypes = new HashMap<>();

    @SuppressWarnings("UnstableApiUsage")
    public MessageSender() {
        if (instance == null)
            instance = this;

        try {
            ClassPath.from(XenLib.getPlugin().getClass().getClassLoader())
                    .getAllClasses()
                    .parallelStream()
                    .filter(clazz -> clazz.getPackageName().equalsIgnoreCase("net.insprill.cjm.messages.types"))
                    .map(ClassPath.ClassInfo::load)
                    .filter(MessageType.class::isAssignableFrom)
                    .filter(c -> c != MessageType.class)
                    .map(clazz -> {
                        try {
                            return clazz.getConstructor().newInstance();
                        } catch (ReflectiveOperationException e) {
                            e.printStackTrace();
                        }
                        return null;
                    })
                    .filter(Objects::nonNull)
                    .forEach(message -> {
                        MessageType messageType = (MessageType) message;
                        messageTypes.put(messageType.getName(), messageType);
                        CJM.getInstance().getMetrics().addCustomChart(new SimplePie("message_type_" + messageType.getName(), () -> {
                            return messageType.isEnabled() + "";
                        }));
                    });
        } catch (IOException e) {
            e.printStackTrace();
        }

        InitPermissions();
    }

    private final List<String> permissions = new ArrayList<>();

    /**
     * Registers all message permissions.
     */
    public void InitPermissions() {
        for (String permission : permissions) {
            XenUtils.unregisterPermission(permission);
        }
        permissions.clear();
        for (MessageType msg : messageTypes.values()) {
            if (!msg.isEnabled())
                continue;
            for (MessageAction action : MessageAction.values()) {
                for (MessageVisibility visibility : MessageVisibility.values()) {
                    String path = visibility.getConfigSection() + "." + action.getConfigSection();
                    for (String key : msg.getConfig().getKeys(path)) {
                        String permission = msg.getConfig().getString(path + "." + key + ".Permission");
                        permissions.add(permission);
                        XenUtils.registerPermission(permission, PermissionDefault.FALSE);
                    }
                }
            }
        }
    }

    /**
     * Handles sending messages for a player.
     *
     * @param player      Player who joined/ left.
     * @param action      Action the player performed.
     * @param vanishCheck Whether to check if the player is vanished before sending messages.
     */
    public void sendMessages(Player player, MessageAction action, boolean vanishCheck) {
        if (action == MessageAction.QUIT && !PlayerHandler.isPlayerLoggedIn(player.getUniqueId()))
            return;
        if (vanishCheck && VanishHandler.isVanished(player))
            return;
        if (!YamlFile.CONFIG.getBoolean("Addons.Jail.Send-Messages-For-Jailed-Players") && PlayerHandler.isJailed(player))
            return;

        for (MessageVisibility visibility : MessageVisibility.values()) {
            if (action == MessageAction.QUIT && visibility == MessageVisibility.PRIVATE)
                continue;

            for (MessageType msg : messageTypes.values()) {
                if (!msg.isEnabled())
                    continue;

                String path = visibility.getConfigSection() + "." + action.getConfigSection();

                // Get the highest priority message the player has access to.
                int hp = msg.getConfig().getKeys(path).stream()
                        .filter(XenMath::isInteger)
                        .filter(num -> PlayerHandler.hasPermission(player, msg.getConfig().getString(path + "." + num + ".Permission")))
                        .mapToInt(Integer::parseInt)
                        .max()
                        .orElse(-1);
                if (hp == -1)
                    continue;

                String messagePath = path + "." + hp;

                int maxPlayers = msg.getConfig().getInt(messagePath + ".Max-Players");
                if (maxPlayers > 0 && Bukkit.getOnlinePlayers().size() > maxPlayers)
                    continue;

                int minPlayers = msg.getConfig().getInt(messagePath + ".Min-Players");
                if (minPlayers > 0 && Bukkit.getOnlinePlayers().size() < minPlayers)
                    return;

                double radius = msg.getConfig().getDouble(messagePath + ".Radius");
                List<Player> players = (visibility == MessageVisibility.PUBLIC)
                        ? PlayerHandler.getNearbyPlayers(player, radius, YamlFile.CONFIG.getBoolean("World-Based"))
                        : Collections.singletonList(player);

                if (action == MessageAction.QUIT && YamlFile.CONFIG.getBoolean("World-Based")) {
                    players.remove(player);
                }

                String randomKey = RandomHandler.getRandomKey(msg.getConfig(), messagePath + "." + msg.getKey());
                if (randomKey == null)
                    continue;

                msg.handle(player, players, path, randomKey, visibility);
            }
        }
    }

}

File: src/main/kotlin/net/insprill/cjm/message/MessageSender.kt
package net.insprill.cjm.message

import de.leonhard.storage.internal.FlatFile
import net.insprill.cjm.CustomJoinMessages
import net.insprill.cjm.extension.getMessage
import net.insprill.cjm.message.types.MessageType
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import org.bukkit.permissions.Permission
import org.bukkit.permissions.PermissionDefault

class MessageSender(private val plugin: CustomJoinMessages) {

    val typeMap = HashMap<String, MessageType>()

    private val registeredPermissions = ArrayList<String>()

    fun registerType(messageType: MessageType) {
        typeMap[messageType.name.lowercase()] = messageType
    }

    fun reloadPermissions(config: FlatFile) {
        registeredPermissions.forEach { Bukkit.getPluginManager().removePermission(it) }
        registeredPermissions.clear()
        for (action in MessageAction.values()) {
            for (visibility in MessageVisibility.values()) {
                val path = visibility.configSection + "." + action.configSection
                for (key in config.singleLayerKeySet(path)) {
                    val permission = config.getString("$path.$key.Permission") ?: continue
                    if (Bukkit.getPluginManager().getPermission(permission) == null && permission != "cjm.default") {
                        Bukkit.getPluginManager().addPermission(Permission(permission, PermissionDefault.FALSE))
                    }
                    registeredPermissions.add(permission)
                }
            }
        }
    }

    /**
     * Handles sending messages for a player. May or may not send messages depending on if all conditions are met.
     *
     * @param player      Player who joined/ left.
     * @param action      Action the player performed.
     * @param vanishCheck Whether to check if the player is vanished before sending messages.
     */
    fun trySendMessages(player: Player, action: MessageAction, vanishCheck: Boolean) {
        if (!action.canRun(plugin, player))
            return
        if (!plugin.toggleHandler.isToggled(player, action))
            return
        if (vanishCheck && plugin.hookManager.isVanished(player))
            return
        if (!plugin.config.getBoolean("Addons.Jail.Ignore-Jailed-Players") && plugin.hookManager.isJailed(player))
            return
        for (visibility in MessageVisibility.values().filter { it.supports(action) }) {
            for (msg in typeMap.values.filter { it.isEnabled }) {
                val path = visibility.configSection + "." + action.configSection

                // Get the highest priority message the player has access to.
                val hp = getHighestPriorityMessage(msg, path, player)
                if (hp == -1)
                    continue

                val messagePath = "$path.$hp"
                if (!MessageCondition.checkAllConditions(msg, messagePath))
                    continue

                val radius = msg.config.getOrDefault("$messagePath.Radius", -1.0)

                val players = getReceivingPlayers(player, visibility, action, radius)
                val randomKey = getRandomKey(msg.config, "$messagePath.${msg.key}") ?: continue
                val delay = msg.config.getLong("$messagePath.Delay")
                val runnable = Runnable { msg.handle(player, players, randomKey, visibility) }
                if (delay < 1) {
                    runnable.run()
                } else {
                    Bukkit.getScheduler().runTaskLater(plugin, runnable, delay)
                }
            }
        }
    }

    private fun getHighestPriorityMessage(msgType: MessageType, path: String, player: Player): Int {
        return msgType.config.singleLayerKeySet(path)
            .mapNotNull { it.toIntOrNull() }
            .filter { player.hasPermission(msgType.config.getOrDefault("$path.$it.Permission", "cjm.default")) }
            .maxOrNull() ?: -1
    }

    private fun getReceivingPlayers(sourcePlayer: Player, visibility: MessageVisibility, action: MessageAction, radius: Double): List<Player> {
        if (visibility != MessageVisibility.PUBLIC) {
            return listOf(sourcePlayer)
        }
        val players = ArrayList(getNearbyPlayers(sourcePlayer, radius, plugin.config.getBoolean("World-Based-Messages.Enabled")))
        if (action == MessageAction.QUIT && plugin.config.getBoolean("World-Based-Messages.Enabled")) {
            players.remove(sourcePlayer)
        }
        return players
    }

    fun getRandomKey(config: FlatFile, path: String): String? {
        val section = config.getSection(path)
        if (section == null || section.singleLayerKeySet().isEmpty()) {
            plugin.logger.severe(plugin.commandManager.getMessage("cjm.message-sender.no-keys", "%path%", path, "%config%", config.file.name))
            return null
        }
        return "$path.${section.singleLayerKeySet().random()}"
    }

    private fun getNearbyPlayers(player: Player, radius: Double, sameWorldOnly: Boolean): List<Player> {
        if (radius > 0) {
            val curr = player.location
            return player.world.players.filter { it.location.distance(curr) <= radius }
        }
        return if (!sameWorldOnly) {
            ArrayList(Bukkit.getOnlinePlayers())
        } else {
            player.world.players
        }
    }

}
